includes/Core/Dashboard_Sharing/Activity_Metrics/Active_Consumers.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\Dashboard_Sharing\Activity_Metrics\Active_Consumers
+ *
+ * @package   Google\Site_Kit\Core\Dashboard_Sharing\Activity_Metrics
+ * @copyright 2022 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Dashboard_Sharing\Activity_Metrics;
+
+use Closure;
+use Google\Site_Kit\Core\Storage\User_Setting;
+
+/**
+ * Class for representing active consumers for an access token.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+class Active_Consumers extends User_Setting {
+
+	/**
+	 * The user option name for this setting.
+	 */
+	const OPTION = 'googlesitekit_active_consumers';
+
+	/**
+	 * Gets the expected value type.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string The type name.
+	 */
+	protected function get_type() {
+		return 'array';
+	}
+
+	/**
+	 * Gets the default value.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array The default value.
+	 */
+	protected function get_default() {
+		return array();
+	}
+
+	/**
+	 * Gets the callback for sanitizing the setting's value before saving.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return Closure
+	 */
+	protected function get_sanitize_callback() {
+		return function ( $value ) {
+			// If the new value is not an array, preserve current value.
+			if ( ! is_array( $value ) ) {
+				return $this->get();
+			}
+
+			// If the new value is not an associative array, preserve current value.
+			if ( array_values( $value ) === $value ) {
+				return $this->get();
+			}
+
+			// If any of the array keys isn't an integer, preserve current value.
+			if ( count(
+				array_filter(
+					array_keys( $value ),
+					function( $item ) {
+						return ! is_int( $item );
+					}
+				)
+			) > 0 ) {
+				return $this->get();
+			}
+
+			// If any of the array values isn't an array, or any of the values of
+			// the nested arrays isn't a string, preserve current value.
+			if ( count(
+				array_filter(
+					array_values( $value ),
```

Comment:
Instead of returning the current value if something set incorrectly, we should omit corrupted values and return only valid ones. Let's say we have this incoming value:

```
array(
    54 => array( 'administrator' ),
    true,
    52 => '',
    11 => 0,
    79 => array( 'editor', 'contributor' ),
)
```

Then the sanitization callback should return just:

```
array(
    54 => array( 'administrator' ),
    79 => array( 'editor', 'contributor' ),
)
```

cc @techanvil do I understand it correctly?

Reply:
Hi @eugene-manuilov, yes, that is the correct shape for the data.

However - I would question whether we do need this behaviour in the `sanitize_callback`. Considering the code which uses `Active_Consumers` is under our control and the data is not coming from some external source which _needs_ to be sanitized, I think that we might actually hide potential bugs by silently reverting to the current value, or indeed by sanitizing data which has no reason to be bad in the first place.

I'd suggest that we should instead simply throw an error if the data is not as expected. An alternative could be to log an error while keeping this logic. But I don't really see why we should ever expect to have to sanitize the data like this...

Reply:
@techanvil 

> I would question whether we do need this behaviour in the `sanitize_callback`

That's the sanitization functions job to remove malicious or corrupted data from the incoming values and return only expected data.

> I think that we might actually hide potential bugs by silently reverting to the current value, or indeed by sanitizing data which has no reason to be bad in the first place.

See the `sanitize_text_field` function from the WP core, it also "silently" removes line breaks, tabs, extra whitespace, etc. 

> Considering the code which uses `Active_Consumers` is under our control and the data is not coming from some external source ... I'd suggest that we should instead simply throw an error if the data is not as expected.

IMO throwing exceptions is against WP approach of handling error states although we use it in a few places. I would avoid using this approach even if we control the codebase.

> But I don't really see why we should ever expect to have to sanitize the data like this...

That's how we do things in our plugin. Check how we sanitize user roles for shared settings, we also remove incorrect roles when shared dashboard settings are saved. So there is nothing new in this sanitize callback function.

Reply:
Thanks for the detail there, @eugene-manuilov.

> See the `sanitize_text_field` function from the WP core, it also "silently" removes line breaks, tabs, extra whitespace, etc.

That makes sense and is a bit of a different use case as it's designed to sanitize user input whereas what we're doing here is sanitizing programmatic input.

However, for the other reasons you have mentioned, I am happy to retract my objection. :)